HPCC-15523 Support JOIN transform MATCHED(LEFT) and MATCHED(RIGHT)

afishbeck opened the pull request:
Enable MATCHED(LEFT) and MATCHED(RIGHT) to be used within the transform
of a JOIN or DENORMALIZE to detect whether the RIGHT or LEFT were matched
rather than RIGHT or LEFT being default records.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

afishbeck commented:
See changes made to the file testing/regress/ecl/genjoin.ecl for examples of usage.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 38844a64e23d1e1fba8dbb9c3cfbe6b2688b0c0f
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 41.227 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |9 |9 |0 |
|setup (thor) |9 |9 |0 |
|setup (roxie) |9 |9 |0 |
|test (hthor) |727 |727 |0 |
|test (thor) |623 |623 |0 |
|test (roxie) |753 |753 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 15e04fbbbf3eddd47ccf3a054ca6daea022f811b
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 42.407 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |9 |9 |0 |
|setup (thor) |9 |9 |0 |
|setup (roxie) |9 |9 |0 |
|test (hthor) |727 |726 |1 |
|test (thor) |623 |623 |0 |
|test (roxie) |753 |753 |0 |


Errors:
- test
  - hthor
    - persist_refresh (persistRefresh=false), 

HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
@ghalliday Please review

AttilaVamos commented:
The persist_refresh (persistRefresh=false) fixed on 6.4.0 and waiting to up-merge to master.

afishbeck commented on rtl/include/eclhelper_base.hpp:
```diff
@@ -1816,9 +1816,8 @@ class CThorJoinArg : implements IHThorJoinArg, public CThorArg
     virtual size32_t onFailTransform(ARowBuilder & rowBuilder, const void * _left, const void * _right, IException * e) { return 0; }
 
 //Join:
-    virtual size32_t transform(ARowBuilder & rowBuilder, const void * _left, const void * _right) { return 0; }
```
@ghalliday FYI as far as I can tell these functions were made common with the denormalize transform below and were never called anymore... but please verify.

ghalliday replied:
Quite possibly

ghalliday commented on ecl/hqlcpp/hqlcpp.cpp:
```diff
@@ -9562,6 +9568,59 @@ void HqlCppTranslator::doBuildExprFailCode(BuildCtx & ctx, IHqlExpression * expr
+    }
+    __int64 mask = 0;
+    IHqlExpression * patternExpr = expr->queryChild(0);
+    if (patternExpr)
```
Possibly an assert instead of an if

ghalliday commented on ecl/hqlcpp/hqlcpp.cpp:
```diff
@@ -9562,6 +9568,59 @@ void HqlCppTranslator::doBuildExprFailCode(BuildCtx & ctx, IHqlExpression * expr
+            throwError(HQLERR_InvalidMatchedPatternInJoin);
+    }
+    IValue * maskValue = createIntValue(mask, 4, false);
+    tgt.expr.setown(createValue(no_band, LINK(matchedMarker->queryExpr()), createConstant(maskValue)));
```
Should include the type of the expression in createValue(), and should probably also compare against 0 to create an expression of boolean type.

ghalliday commented on ecl/hqlcpp/hqlcpp.cpp:
```diff
@@ -9562,6 +9568,59 @@ void HqlCppTranslator::doBuildExprFailCode(BuildCtx & ctx, IHqlExpression * expr
+    tgt.expr.setown(createValue(no_band, LINK(matchedMarker->queryExpr()), createConstant(maskValue)));
+}
+
+void HqlCppTranslator::doBuildAssignMatchedInJoin(BuildCtx & ctx, const CHqlBoundTarget & target, IHqlExpression * expr)
```
There isn't any advantage in special casing assignment, so should probably go through the default processing (and remove this function).  (Eventually I may refactor all this code).

ghalliday commented on ecl/hqlcpp/hqlhtcpp.cpp:
```diff
@@ -17406,6 +17408,12 @@ void HqlCppTranslator::buildSOAPtoXml(BuildCtx & ctx, IHqlExpression * dataset,
+void HqlCppTranslator::associateLocalJoinTransformFlags(BuildCtx & ctx, const char * name)
+{
+    OwnedHqlExpr activeJoinTransformFlagsMarker = createAttribute(activeJoinTransformFlagsAtom);
+    OwnedHqlExpr activeJoinTransformFlagsVariable = createVariable(name, makeBoolType());
```
Strictly speaking the type is unsigned4.

ghalliday reviewed (commented):
@afishbeck looks good.
A few fairly minor comments.

Very obscure, but what happens if a JOIN is converted to a PROJECT (can occur for a left outer join,keep(1))if it contains no_matched_injoin.

I'm slightly torn on whether MATCHED(RIGHT) should also be implemented for the grouped denormalize - you can already probably check by checking the number of rows, but it would provide consistency.

afishbeck commented:
"but what happens if a JOIN is converted to a PROJECT".. I'll try and create that scenario so I can fully understand.  But sounds like I'll have to emulate the MATCHED support there to allow the transform to work consistently.

"grouped denormalize".. Agreed, consistency is good, I'll add support there.

ghalliday commented:
@afishbeck as we discussed at the offsite, code like the following should avoid the problems:

    | MATCHED '(' dataRow ')'
                         {
                             $$.setExpr(createValue(no_matched_injoin, makeBoolType(), $3.getExpr()));

afishbeck commented:
Updated to use MATCHED(row) syntax and to support denormalize group.

@ghalliday please review.

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: f0b633ef12fa2fbf02b900f6c6be896fd06e40bc
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 42.663 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |736 |736 |0 |
|test (thor) |630 |630 |0 |
|test (roxie) |765 |763 |2 |


Errors:
- test
  - roxie
    - genjoin_nothor, schedule2, 

HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: f0b633ef12fa2fbf02b900f6c6be896fd06e40bc
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 46.088 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |736 |736 |0 |
|test (thor) |630 |630 |0 |
|test (roxie) |765 |763 |2 |


Errors:
- test
  - roxie
    - genjoin_nothor, schedule2, 

HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
@afishbeck What's up with the genjoin fialure on Roxie?

afishbeck commented:
@richardkchapman it's not a regression, I've added to the test case for my new functionality but looks like I messed up a flag for one roxie case at the last minute.. verifying.

afishbeck commented:
@richardkchapman updated

@ghalliday please review

richardkchapman commented on testing/regress/ecl/genjoin.ecl:
```diff
@@ -25,8 +25,10 @@ jrec := RECORD,MAXLENGTH(100)
     UNSIGNED1 i;
     STRING3 lstr;
     STRING3 rstr;
-    UNSIGNED1 c;
```
See line 18:

 // IMPORTANT: this test is generated by the perl script genjointest.pl, so do not edit it by hand

afishbeck replied:
oops... I actually saw that but since the file is checked in to github thought it must have been a historical remnant.  Should I change the script, or create separate versions for testing this new functionality?

afishbeck replied:
I need full coverage of various types of join and denormalize to make sure the match flags are being set correctly.

richardkchapman replied:
I don't know whether it's a historical remnant or not... Try running genjointest.pl and see how different the generated tests are from what is checked in.

richardkchapman replied:
(it's ok to check in the resulting test files but should also check in updated genjointest.pl if it IS still valid...)

afishbeck replied:
sure, will do.

afishbeck commented on ecl/hqlcpp/hqlcpp.cpp:
```diff
@@ -9607,6 +9610,61 @@ void HqlCppTranslator::doBuildExprFailCode(BuildCtx & ctx, IHqlExpression * expr
+    node_operator op = matchExpr->getOperator();
+    if (op==no_left)  //included for completeness, always return true
+    {
+        tgt.expr.setown(createConstant(createBoolValue(true)));
```
@ghalliday I don't think I can make this assumption so may have to pass flags for matching left here as well.  But please go ahead and review the rest.

afishbeck replied:
If I do use flags for left I might as well use them for the right dataSet as well and make everything look the same.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 7869d4e16809017217b13820dc52f72239828fda
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 42.412 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |736 |736 |0 |
|test (thor) |630 |630 |0 |
|test (roxie) |765 |765 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

afishbeck commented:
@ghalliday updated to use flags for group denormalize as well.  Also makes implementation entirely consistent.

Just have to check whether I need to update genjointest.pl script, but that wont affect implementation.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: e3d32af7eca17e0109859444920213cb01fd0c21
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 44.702 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |736 |736 |0 |
|test (thor) |630 |630 |0 |
|test (roxie) |765 |765 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

afishbeck commented:
@richardkchapman I've fixed and updated genjointest.pl.  It had minor syntax errors so probably hasn't been run in a while, but it did still match the checked in ecl files.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 14673d9dc7ba51cdcdf6297d4a3888e756c504f9
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 41.243 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |736 |736 |0 |
|test (thor) |630 |630 |0 |
|test (roxie) |765 |765 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented:
@afishbeck please see various comments above.

afishbeck commented on ecl/hqlcpp/hqlcpp.cpp:
```diff
@@ -9607,6 +9610,41 @@ void HqlCppTranslator::doBuildExprFailCode(BuildCtx & ctx, IHqlExpression * expr
+        return;
+    }
+
+    __int64 mask = 0;
```
@ghalliday I think if you could help create a test case that would help more than anything.  There is something about matching the correct layer of nested joins right and left that is probably quite straight forward but that I'm just not yet picturing fully.

afishbeck replied:
I do understand why you want it associated in associateLocalJoinTransformFlags, just something in the details I need to tie it all together.

afishbeck commented on rtl/include/eclhelper.hpp:
```diff
@@ -1743,9 +1749,9 @@ struct IHThorAnyJoinBaseArg : public IHThorArg
 //Join:
 //Denormalize
-    virtual size32_t transform(ARowBuilder & rowBuilder, const void * _left, const void * _right, unsigned _count) { return 0; }
+    virtual size32_t transform(ARowBuilder & rowBuilder, const void * _left, const void * _right, unsigned _count, unsigned _flags) { return 0; }
```
Will do. I thought the numbers were already increased for 7.0.

afishbeck commented on ecl/hql/hqlgram.y:
```diff
@@ -6495,6 +6495,14 @@ primexpr1
+                        {
+                            $$.setExpr(createValue(no_matched_injoin, makeBoolType(), $3.getExpr()));
+                        }
+    | MATCHED '(' dataSet ')'
```
I may have just assumed.  I'll try without this.

afishbeck replied:
MATCHED(ROWS(RIGHT))
@ghalliday removing this does cause a syntax error trying to MATCH(right) on a group denormalize dataset.  Not sure if there is something else I'm missing.

afishbeck commented on ecl/hqlcpp/hqlhtcpp.cpp:
```diff
@@ -17464,6 +17468,11 @@ void HqlCppTranslator::buildSOAPtoXml(BuildCtx & ctx, IHqlExpression * dataset,
     buildXmlSerialize(func.ctx, transform->queryRecord(), self, &assigns);
 }
 
+void HqlCppTranslator::associateLocalJoinTransformFlags(BuildCtx & ctx, const char * name)
```
Ah I think something may have just clicked, thanks.

afishbeck commented on thorlcr/activities/lookupjoin/thlookupjoinslave.cpp:
```diff
@@ -1188,7 +1188,7 @@ class CInMemJoinBase : public CSlaveActivity, public CAllOrLookupHelper<HELPER>,
             if (isGroupOp())
             {
-                size32_t sz = HELPERBASE::joinTransform(rowBuilder, leftRow, rightRow, numRows, filteredRhs.getArray());
+                size32_t sz = HELPERBASE::joinTransform(rowBuilder, leftRow, rightRow, numRows, filteredRhs.getArray(), JTFmatchedleft|JTFmatchedright);
```
The if at the top of this block checks "if (filteredRhs.ordinality()"
I think the line:
            const void *rightRow = numRows ? filteredRhs.item(0) : defaultRight.get();
could also be simplified.

afishbeck replied:
yes, fixed

afishbeck commented on testing/regress/ecl/nestedjoin.ecl:
```diff
@@ -0,0 +1,90 @@
+BOOLEAN match(jrec l, jrec r) := (match1(l, r) AND match2(l, r));
+
+// transform for joins and non-group denormalizes, to be used with match
+childrec xfmchild(jrec lparent, childrec l, jrec rparent, childrec r, STRING lab) := TRANSFORM
```
@ghalliday Passing in the parent records here is the only way I found to cause my previous implementation to behave incorrectly.  If I were to have done MATCHED(lparent) here my old version would have acted as if I did MATCHED(l).

Now it would (correctly I think) give an error saying you can only match on the local left or write.

My old implementation worked fine as far as I could tell if you weren't passing in the parent like that.

afishbeck commented:
@ghalliday updated, please review.  I'll squash the commits after you've reviewed the latest.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: e5668eee646f211eca6974c01521c86208652002
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 43.856 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |737 |737 |0 |
|test (thor) |631 |631 |0 |
|test (roxie) |766 |766 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented on testing/regress/ecl/nestedjoin.ecl:
```diff
@@ -0,0 +1,90 @@
+    SELF.ml := MATCHED(l);
+    SELF.mr := MATCHED(r);
+    SELF.vr := IF (MATCHED(r), -1, 1);
+END;
```
You should also check the results of MATCHED(lParent), MATCHED(rParent) to ensure they are correct.

Although highly unusual things like this do occur.  A typical situation is where an attribute is defined in a transform based on MATCHED(LEFT).  That value is then passed into a function which happens to contain a join which is done as a child query, and the value that was passed in is used in that function's transform.

ghalliday reviewed (commented):
Great, it is looking very close now.   If you update the test as mentioned I will re-review the whole PR.

(I am slightly paranoid because if results are wrong it is unlikely to get spotted.)

ghalliday commented:
@afishbeck one comment on the test.

afishbeck commented:
@ghalliday updated with MATCH(lparent) enabled, working, and in the test case.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: ec7b43a97860070ad2c6c28c74d0e14de9f3d689
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 44.845 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |737 |737 |0 |
|test (thor) |631 |631 |0 |
|test (roxie) |766 |766 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented:
Double checked hthor, roxie and thor and all the calls from the engines look correct.

ghalliday commented on ecl/hql/hqlgram.y:
```diff
@@ -6495,6 +6495,14 @@ primexpr1
                         }
+    | MATCHED '(' dataRow ')'
+                        {
+                            $$.setExpr(createValue(no_matched_injoin, makeBoolType(), $3.getExpr()));
```
trivial: Should add an extra $1 parameter to the setExpr to record the source code position.

ghalliday reviewed (commented):
@afishbeck a few more comments

afishbeck commented on testing/regress/ecl/genjoin.ecl:
```diff
@@ -67,6 +76,9 @@ jrec xfmgrp(jrec l, DATASET(jrec) r, STRING lab) := TRANSFORM
     SELF.rstr := r[c].rstr;
     SELF.c := c;
     SELF.label := lab;
+    SELF.ml := MATCHED(l);
```
genjoin_nothor.ecl and genjoin_justroxie.ecl have coverage of group denormalize, not sure if it's complete or not.

afishbeck replied:
all three test files are generated from genjointest.pl with different selections for test cases.

afishbeck replied:
but I think that means it would never get tested in THOR.

afishbeck commented:
@ghalliday updated

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 0a748161643f79bdbd88ae7facc05625d318a589
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 42.85 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |741 |0 |
|test (thor) |636 |636 |0 |
|test (roxie) |768 |768 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented:
@afishbeck please squash - it looks ready to merge.

afishbeck committed:
HPCC-15523 Support JOIN transform MATCHED(LEFT) and MATCHED(RIGHT)

Enable MATCHED(LEFT) and MATCHED(RIGHT) to be used within the transform
of a JOIN or DENORMALIZE to detect whether the RIGHT or LEFT was matched
rather than RIGHT or LEFT being default records.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

afishbeck commented:
@ghalliday squashed and ready to merge

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 77ac56a839d8365f11cac5be9aebbc79090d2460
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 44.413 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |741 |0 |
|test (thor) |636 |636 |0 |
|test (roxie) |768 |768 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
